sql/sql_acl.cc

```diff
@@ -3173,18 +3144,39 @@ check_access(THD *thd, privilege_t want_access,
     DBUG_RETURN(FALSE);
   }
 
+  /*
+    Shortcut to not check database specific access rights. One caveat:
+      If we don't have a global SELECT privilege, we have to get the
+      database specific access rights to be able to handle queries of type
+      UPDATE t1 SET a=1 WHERE b > 0
+      In that case the shortcut won't work.
```

Comment:
the original code didn't make any sense to me, so the refactoring preserving the same logic didn't help much :)
a better comment for this 20-year old code could've been
```c++
/*
  for UPDATE and DELETE the server only checks for the UPDATE_ACL
  (or DELETE_ACL) first, so want_access is UPDATE_ACL or
  DELETE_ACL here. later it'll check for SELECT_ACL if there old
  column values are used in the statement. to simplify this later
  check let's merge global and database privileges in *saved_priv
  now when global privileges alone don't include SELECT_ACL
*/
```
I think it's not a nice solution at all, more like a hack. Instead db privileges have to be consulted when they're actually needed (during the check for SELECT_ACL).